File: test/osgx_Shapes.py
import osgx

from OpenSceneGraph import *

def test_named_shapes_have_expected_topology_and_face_basis():
	# These counts are the unexpanded polyhedron data exposed to a dice layer.  Geometry itself
	# deliberately expands faces into flat-shaded triangle vertices during rebuild().
	expected = {
		osgx.Cube: (8, 6),
		osgx.Tetrahedron: (4, 4),
		osgx.Octahedron: (6, 8),
		osgx.Icosahedron: (12, 20),
		osgx.Dodecahedron: (20, 12),
		osgx.PentagonalTrapezohedron: (12, 10),
	}

	for kind, (vertex_count, face_count) in expected.items():
		shape = kind()

		assert len(shape.vertices) == vertex_count
		assert len(shape.faces) == face_count

		normal = shape.faceNormal(0)
		up = shape.faceUp(0)

		assert abs(normal.length() - 1.0) < 1e-5
		assert abs(up.length() - 1.0) < 1e-5
		assert abs(normal.dot(up)) < 1e-5
		assert shape.restingOffset() >= 0.0

def test_named_shapes_accept_a_common_circumradius():
	for kind in (
		osgx.Cube,
		osgx.Tetrahedron,
		osgx.Octahedron,
		osgx.Icosahedron,
		osgx.Dodecahedron,
		osgx.PentagonalTrapezohedron,
	):
		shape = kind(radius=2.0)

		assert abs(max(vertex.length() for vertex in shape.vertices) - 2.0) < 1e-5

def test_face_resting_offset_places_the_selected_face_on_the_floor():
	cube = osgx.Cube(size=osg.Vec3(2.0, 2.0, 2.0))

	# Every cube face lies one unit from its center.  This specifically tests
	# the selected support plane, rather than the lowest vertex used by
	# restingOffset().
	for face_index, face in enumerate(cube.faces):
		orientation = osg.Quat()
		orientation.makeRotate(cube.faceNormal(face_index), osg.Vec3(0.0, 0.0, -1.0))
		offset = cube.faceRestingOffset(face_index, orientation)

		for vertex_index in face.vertices:
			assert abs((orientation * cube.vertices[vertex_index]).z + offset) < 1e-5

def test_face_derives_a_stable_local_plane_from_polyhedron_positions():
	cube = osgx.Cube(size=osg.Vec3(2.0, 2.0, 2.0))
	face = cube.faces[0] # -Z: 0, 3, 2, 1

	assert face.origin(cube.vertices) == osg.Vec3(-1.0, -1.0, -1.0)
	assert face.center(cube.vertices) == osg.Vec3(0.0, 0.0, -1.0)
	assert face.normal(cube.vertices) == osg.Vec3(0.0, 0.0, -1.0)
	assert face.right(cube.vertices) == osg.Vec3(0.0, 1.0, 0.0)
	assert face.up(cube.vertices) == osg.Vec3(1.0, 0.0, 0.0)
	assert face.planeCoordinates(cube.vertices) == [
		osg.Vec2(0.0, 0.0), osg.Vec2(2.0, 0.0),
		osg.Vec2(2.0, 2.0), osg.Vec2(0.0, 2.0)
	]

def test_named_shape_faces_reconstruct_from_their_local_planes():
	for kind in (
		osgx.Cube,
		osgx.Tetrahedron,
		osgx.Octahedron,
		osgx.Icosahedron,
		osgx.Dodecahedron,
		osgx.PentagonalTrapezohedron,
	):
		shape = kind()

		for face in shape.faces:
			origin = face.origin(shape.vertices)
			right = face.right(shape.vertices)
			up = face.up(shape.vertices)
			normal = face.normal(shape.vertices)

			for vertex_index, coordinate in zip(face.vertices, face.planeCoordinates(shape.vertices)):
				reconstructed = origin + right * coordinate.x + up * coordinate.y

				assert (reconstructed - shape.vertices[vertex_index]).length() < 1e-5

			assert abs((face.center(shape.vertices) - origin).dot(normal)) < 1e-5

def test_polyhedron_expands_face_attributes_and_removes_them():
	face = osgx.Polyhedron.Face(
		[0, 1, 2],
		[osg.Vec2(0.0, 0.0), osg.Vec2(1.0, 0.0), osg.Vec2(0.0, 1.0)]
	)
	polyhedron = osgx.Polyhedron(
		[osg.Vec3(0.0, 0.0, 0.0), osg.Vec3(1.0, 0.0, 0.0), osg.Vec3(0.0, 1.0, 0.0)],
		[face]
	)

	# One value per source face becomes one value for every generated triangle vertex.
	polyhedron.setFaceAttribute(13, osg.Vec3Array([osg.Vec3(7.0, 8.0, 9.0)]))

	assert len(polyhedron.vertexAttrib[13]) == 3

	polyhedron.removeAttribute(13)

	assert 13 not in polyhedron.vertexAttrib

File: ext/python/osgx-shapes.cpp
#include "osgx-python.hpp"
#include "osgx/Shapes.hpp"

namespace osgx_python {

void bind_shapes(py::module_& m) {
	py::class_<osgx::VertexLayout>(m, "VertexLayout")
		.def(py::init<>())
		.def_readwrite("position", &osgx::VertexLayout::position)
		.def_readwrite("normal", &osgx::VertexLayout::normal)
		.def_readwrite("uv", &osgx::VertexLayout::uv)
	;

	py::class_<
		osgx::Polyhedron,
		osg::Geometry,
		osg::ref_ptr<osgx::Polyhedron>
	> polyhedron(m, "Polyhedron");

	py::class_<osgx::Polyhedron::Face>(polyhedron, "Face")
		.def(py::init<>())
		.def(
			py::init<std::vector<std::uint32_t>, std::vector<osg::Vec2>>(),
			"vertices"_a,
			"uv"_a=std::vector<osg::Vec2>{}
		)
		.def_readwrite("vertices", &osgx::Polyhedron::Face::vertices)
		.def_readwrite("uv", &osgx::Polyhedron::Face::uv)
		.def("origin", &osgx::Polyhedron::Face::origin, "positions"_a)
		.def("center", &osgx::Polyhedron::Face::center, "positions"_a)
		.def("normal", &osgx::Polyhedron::Face::normal, "positions"_a)
		.def("right", &osgx::Polyhedron::Face::right, "positions"_a)
		.def("up", &osgx::Polyhedron::Face::up, "positions"_a)
		.def("planeCoordinates", &osgx::Polyhedron::Face::planeCoordinates, "positions"_a)
	;

	polyhedron
		.def(py::init<>())
		.def(
			py::init<
				std::vector<osg::Vec3>,
				std::vector<osgx::Polyhedron::Face>,
				osgx::VertexLayout
			>(),
			"vertices"_a,
			"faces"_a,
			"layout"_a=osgx::VertexLayout{}
		)
		.def_property(
			"vertices",
			[](const osgx::Polyhedron& self) { return self.vertices(); },
			&osgx::Polyhedron::setVertices
		)
		.def_property(
			"faces",
			[](const osgx::Polyhedron& self) { return self.faces(); },
			&osgx::Polyhedron::setFaces
		)
		.def_property("layout", &osgx::Polyhedron::layout, &osgx::Polyhedron::setLayout)
		.def("setFaceVertexAttribute", &osgx::Polyhedron::setFaceVertexAttribute)
		.def("setFaceAttribute", &osgx::Polyhedron::setFaceAttribute)
		.def("removeAttribute", &osgx::Polyhedron::removeAttribute)
		.def("rebuild", &osgx::Polyhedron::rebuild)
		.def("faceNormal", py::overload_cast<std::size_t>(&osgx::Polyhedron::faceNormal, py::const_))
		.def("faceUp", &osgx::Polyhedron::faceUp)
		.def("restingOffset", &osgx::Polyhedron::restingOffset, "orientation"_a=osg::Quat())
		.def(
			"faceRestingOffset",
			&osgx::Polyhedron::faceRestingOffset,
			"faceIndex"_a,
			"orientation"_a=osg::Quat()
		)
		.def_static("isometricFaceUV", &osgx::Polyhedron::isometricFaceUV)
	;

	py::class_<
		osgx::Cube,
		osgx::Polyhedron,
		osg::ref_ptr<osgx::Cube>
	>(m, "Cube")
		.def(
			py::init<const osg::Vec3&, const osg::Vec3&, osgx::VertexLayout>(),
			"center"_a=osg::Vec3(),
			"size"_a=osg::Vec3(1.0f, 1.0f, 1.0f),
			"layout"_a=osgx::VertexLayout{}
		)
		.def(
			py::init<const osg::Vec3&, float, osgx::VertexLayout>(),
			"center"_a=osg::Vec3(),
			"radius"_a=1.0f,
			"layout"_a=osgx::VertexLayout{}
		)
	;

	py::class_<
		osgx::Tetrahedron,
		osgx::Polyhedron,
		osg::ref_ptr<osgx::Tetrahedron>
	>(m, "Tetrahedron")
		.def(
			py::init<const osg::Vec3&, float, osgx::VertexLayout>(),
			"center"_a=osg::Vec3(), "radius"_a=1.0f, "layout"_a=osgx::VertexLayout{}
		)
	;

	py::class_<
		osgx::Octahedron,
		osgx::Polyhedron,
		osg::ref_ptr<osgx::Octahedron>
	>(m, "Octahedron")
		.def(
			py::init<const osg::Vec3&, float, osgx::VertexLayout>(),
			"center"_a=osg::Vec3(), "radius"_a=1.0f, "layout"_a=osgx::VertexLayout{}
		)
	;

	py::class_<
		osgx::Icosahedron,
		osgx::Polyhedron,
		osg::ref_ptr<osgx::Icosahedron>
	>(m, "Icosahedron")
		.def(
			py::init<const osg::Vec3&, float, osgx::VertexLayout>(),
			"center"_a=osg::Vec3(), "radius"_a=1.0f, "layout"_a=osgx::VertexLayout{}
		)
	;

	py::class_<
		osgx::Dodecahedron,
		osgx::Polyhedron,
		osg::ref_ptr<osgx::Dodecahedron>
	>(m, "Dodecahedron")
		.def(
			py::init<const osg::Vec3&, float, osgx::VertexLayout>(),
			"center"_a=osg::Vec3(), "radius"_a=1.0f, "layout"_a=osgx::VertexLayout{}
		)
	;

	py::class_<
		osgx::PentagonalTrapezohedron,
		osgx::Polyhedron,
		osg::ref_ptr<osgx::PentagonalTrapezohedron>
	>(m, "PentagonalTrapezohedron")
		.def(
			py::init<const osg::Vec3&, float, osgx::VertexLayout>(),
			"center"_a=osg::Vec3(), "radius"_a=1.0f, "layout"_a=osgx::VertexLayout{}
		)
	;
}

}
